File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        Arrays.sort(times);
        return find(n, times);
    }
    
    static long find(int people, int[] times){
        long start = 1, end = (long)people * times[times.length - 1] + 1;
        
        while(start < end){
			long mid = (start + end) >> 1;
            long possibleCnt = 0;
            for(int i = 0; i < times.length; i++){
                possibleCnt += mid / times[i];
            }
            
            if(possibleCnt >= people){
                end = mid;
                continue;
            }
            start = mid + 1;
        }
        
        return end;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    times = sorted(times)
    lo, hi = 0, times[-1] * n + 1
    while lo + 1 < hi:
        mid = (lo + hi) >> 1
        
        cnt = sum([mid // time for time in times])
        
        if cnt >= n:
            hi = mid
            continue
        lo = mid
    return hi